bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java

```diff
@@ -1091,24 +1156,29 @@ private Header getHeaderForLogId(long entryLogId) throws IOException {
     }
 
     private BufferedReadChannel getChannelForLogId(long entryLogId) throws IOException {
-        BufferedReadChannel fc = getFromChannels(entryLogId);
-        if (fc != null) {
-            return fc;
+        BufferedReadChannel brc = getFromChannels(entryLogId);
+        if (brc != null) {
+            return brc;
         }
+
         File file = findFile(entryLogId);
         // get channel is used to open an existing entry log file
         // it would be better to open using read mode
         FileChannel newFc = new RandomAccessFile(file, "r").getChannel();
-        FileChannel oldFc = logid2FileChannel.putIfAbsent(entryLogId, newFc);
+        ReferenceCountedFileChannel oldFc =
+                logid2FileChannel.putIfAbsent(entryLogId, new ReferenceCountedFileChannel(newFc));
         if (null != oldFc) {
             newFc.close();
-            newFc = oldFc;
+            newFc = oldFc.fc;
+            // increment the refCnt
+            oldFc.retain();
```

Comment:
It's possible that oldFc could have been released between getting it from logid2FileChannel and calling retain, so the channel returned could be closed.

Reply:
you're right, I'll consider this later, thanks.